### DISCUSSION TITLE
Is there any reason why neorv32_cpu_delay_us does not exist?

### DISCUSSION THREAD

**Discussion by agamez** [stnolting/neorv32#498] [Q&A] [answered]
Milliseconds is quite big a delay, sometimes if we want to bit-bang GPIO for any hardware-controlling issue we need lower sleep times.

Isn't it possible to write neorv32_cpu_delay_us just like neorv32_cpu_delay_ms just by simply replacing

`clock = clock / 1000`

with

`clock = clock / 1000000`

Or is there something that I'm not seeing?

**Comment by stnolting** [comment, marked as answer]
> Milliseconds is quite big a delay, sometimes if we want to bit-bang GPIO for any hardware-controlling issue we need lower sleep times.

Good point.

> Isn't it possible to write neorv32_cpu_delay_us just like neorv32_cpu_delay_ms just by simply replacing [...]

Sure, that would be possible. There is just one problem here: if the processor is running with less than 1000000 Hz then `clock = clock / 1000000` would result zero (integer-only arithmetic). Furthermore, with a division of 1000000 you loose "more" precision as you would when dividing by 1000 making the delay routine (and thus, the timing) less precise.
